integrations/uptrain/src/uptrain_haystack/evaluator.py

```diff
@@ -0,0 +1,194 @@
+import json
+import os
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
+
+from haystack import DeserializationError, component, default_from_dict, default_to_dict
+from uptrain import APIClient, EvalLLM, Evals
+from uptrain.framework.evals import ParametricEval
+
+from uptrain_haystack.metrics import (
+    METRIC_DESCRIPTORS,
+    InputConverters,
+    OutputConverters,
+    UpTrainMetric,
+    UpTrainMetricResult,
+)
+
+
+@component
+class UpTrainEvaluator:
+    """
+    A component that uses the UpTrain framework to evaluate inputs against a specific metric.
+
+    The supported metrics are defined by :class:`UpTrainMetric`. The inputs of the component
+    metric-dependent. The output is a list of :class:`UpTrainEvaluatorOutput` objects, each
+    containing a single input and the result of the evaluation performed on it.
+    """
+
+    _backend_metric: Union[Evals, ParametricEval]
+    _backend_client: Union[APIClient, EvalLLM]
+
+    def __init__(
+        self,
+        metric: Union[str, UpTrainMetric],
+        metric_params: Optional[Dict[str, Any]] = None,
+        *,
+        api: str = "openai",
+        api_key_env_var: Optional[str] = "OPENAI_API_KEY",
+        api_params: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Construct a new UpTrain evaluator.
+
+        :param metric:
+            The metric to use for evaluation.
+        :param metric_params:
+            Parameters to pass to the metric's constructor.
+        :param api:
+            The API to use for evaluation.
+
+            Supported APIs: "openai", "uptrain".
+        :param api_key_env_var:
+            The name of the environment variable containing the API key.
+        :param api_params:
+            Additional parameters to pass to the API client.
+        """
+        self.metric = metric if isinstance(metric, UpTrainMetric) else UpTrainMetric.from_str(metric)
+        self.metric_params = metric_params
+        self.descriptor = METRIC_DESCRIPTORS[self.metric]
+        self.api = api
+        self.api_key_env_var = api_key_env_var
+        self.api_params = api_params
+
+        self._init_backend()
+        expected_inputs = self.descriptor.input_parameters
+        component.set_input_types(self, **expected_inputs)
+
+    @component.output_types(output=List["UpTrainEvaluatorOutput"])
+    def run(self, **inputs) -> Dict[str, Any]:
+        """
+        Run the UpTrain evaluator.
+
+        :param inputs:
+            The inputs to evaluate. Match the input parameters of the metric.
+        :returns:
+            A list of :class:`UpTrainEvaluatorOutput` objects, each containing
```

Comment:
As discussed, the run method should better run on a single input sample and return a single output sample rather than lists of input and output samples. 
The list processing I would expect in a what we called `run_batch` method in the past. It's best to sync here with team members working on the pipeline improvements. @silvanocerza 
My last status was that each component should have a run method and optionally they can define a "run_batch" method that takes a list of each of the inputs of the run method.

Reply:
Since this is an ongoing discussion, we'll continue with the current implementation for the time being.